app/assets/javascripts/discourse/routes/unknown.js

```diff
@@ -1,8 +1,29 @@
 import DiscourseRoute from "discourse/routes/discourse";
+import RSVP from "rsvp";
 import { ajax } from "discourse/lib/ajax";
 
 export default DiscourseRoute.extend({
-  model() {
-    return ajax("/404-body", { dataType: "html" });
+  model(params) {
+    const path = params.path;
+    const bodyReq = ajax("/404-body", { dataType: "html" });
+    const testReq = ajax("/permalink-check.json", {
+      data: { path }
+    });
+    return RSVP.allSettled([bodyReq, testReq]).then(results => {
+      const [bodyResp, testResp] = results;
+      if (testResp.state === "fulfilled" && testResp.value) {
+        if (testResp.value.internal) {
+          this.transitionTo(testResp.value.target_url);
+        } else {
+          window.location = testResp.value.target_url;
```

Comment:
I think it's safer to use: https://github.com/discourse/discourse/blob/master/app/assets/javascripts/discourse/lib/url.js#L303